### PR TITLE
#5851 Fix AttributeError on t.i.ssl.Certificate repr

### LINKED ISSUE #5851: t.i.ssl.Certificate __repr__ assumes certificate will have a common name.
|mmcc's avatar| mmcc reported|
|-|-|
|Trac ID|trac#5851|
|Type|defect|
|Created|2012-08-06 22:29:24Z|

It raises AttributeError when a cert has no common name.

```python
>>> from twisted.internet import ssl
>>>pemfn = 'UbuntuOne-Go_Daddy_CA.pem'
>>> pemd = open(pemfn, 'r').read()
>>> ssl.Certificate.loadPEM(pemd)
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "Twisted-11.1.0-py2.7-macosx-10.7-x86_64.egg/twisted/internet/_sslverify.py", line 166, in __repr__
    self.getIssuer().commonName)
  File "Twisted-11.1.0-py2.7-macosx-10.7-x86_64.egg/twisted/internet/_sslverify.py", line 85, in __getattr__
    raise AttributeError(attr)
AttributeError: commonName
>>> 
```

<details><summary>Searchable metadata</summary>

```
trac-id__5851 5851
type__defect defect
reporter__mmcc mmcc
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__new new
resolution__None None
component__core core
keywords__ 
time__1344292164000000 1344292164000000
changetime__1344339159000000 1344339159000000
version__None None
owner__ 

```
</details>

### PR DESCRIPTION
**Pull request by rvandam**
## Scope and purpose

Fixes #5851

### CODE REVIEW

**Comment by adiroiban** [comment]
Many thanks, Rob for the fix.

For a  PR to be merged the changes need to have a release note fragment and the changes to include the tests.

More info about the release notes here 
https://docs.twisted.org/en/latest/development/dev-process.html#release-notes-management

Let us  know if you need  help with the release notes or writing the tests.

Regards

**Comment by graingert** [comment]
@rvandam I'm keen to get this merged do you need help writing a test for this change?

**Comment by rvandam** [comment]
That would be great. Haven’t had time to get back to it yet.

File: src/twisted/test/test_sslverify.py [hunk 1]
```diff
@@ -1474,6 +1474,63 @@ def test_inspectCertificate(self):
...
+            RE4+uXR21aITVSzGh6O1mawGhId/dQb8vxRMDsxuxN89txJx9OjxUUAiKEngHUuH
+            qDTMBqLdElrRhjZkAzVvb3du6/KFUJheqwNTrZEjYx8WnM25sgVjOuH0aBsXBTWV
+            U+4=
+            -----END CERTIFICATE-----
+            """
+        )
+        c = sslverify.Certificate.loadPEM(ubuntuOneGoDaddyPem)
+        self.assertEqual(
+            repr(c),
+            "<Certificate Subject=b'Go Daddy Secure Certification Authority' Issuer=>",
```
Comment (graingert) [inline src/twisted/test/test_sslverify.py:1533]: might be better to omit the `Issuer=` entirely here?

**Comment by graingert** [comment]
@adiroiban I've added a test case to this PR

File: src/twisted/newsfragments/5851.bugfix [hunk 2]
```diff
@@ -0,0 +1 @@
+fix t.i.ssl.Certificate __repr__ for certificates without a common name.
```
Comment (adiroiban) [inline src/twisted/newsfragments/5851.bugfix]: Maybe have it more like a release announcement and less like a commit messaage :)

```suggestion
twisted.internet.ssl.Certificate.__repr__ can now handle certificates without a common name (CN) in the certificate itself or the signing CA.
```

File: src/twisted/test/test_sslverify.py (hunk 1 above)
Comment (adiroiban) [reply src/twisted/test/test_sslverify.py:1533]: I think this is ok. no need to spend too much time with the repr part. this is only to help with the debugging.

File: src/twisted/test/test_sslverify.py [hunk 3]
```diff
@@ -1474,6 +1474,63 @@ def test_inspectCertificate(self):
             ],
         )
 
+    def test_representationOfCertificate(self):
+        """
+        Test that the repr of L{sslverify.Certificate} returns
+        a human-readable string containing some basic information about the
+        certificate.
```
Comment (adiroiban) [inline src/twisted/test/test_sslverify.py]: Maybe be more specific about the scope/purpose of the test.


```suggestion
        The repr of L{sslverify.Certificate} returns
        a human-readable string containing the subject of the certificate itself, and the subject of the signing CA.
```

File: src/twisted/test/test_sslverify.py [hunk 4]
```diff
@@ -1474,6 +1474,63 @@ def test_inspectCertificate(self):
...
+            repr(c),
+            "<Certificate Subject=b'example.twistedmatrix.com' Issuer=b'example.twistedmatrix.com'>",
+        )
+
+    def test_representationOfCertificateNoCN(self):
+        """
+        Test that the repr of L{sslverify.Certificate} returns
+        a human-readable string containing some basic information about the
+        certificate when the ceritificate has no CN.
+        """
```
Comment (adiroiban) [inline src/twisted/test/test_sslverify.py]:
```suggestion
        """
        Repr of a L{sslverify.Certificate} for which both the certificate and the signing CA have no subject, contains the empty string as the subject.
        """
```

Maybe is better to go with a more barebone certificate,
and a certificate that it's easy to regenerate.

```
-----BEGIN CERTIFICATE-----
MIIBejCB3aADAgECAghy9M9u80Q60DAKBggqhkjOPQQDAjAAMB4XDTIzMDkxMTE0
MjIwMFoXDTM0MDkxMTE0MjIwMFowADCBmzAQBgcqhkjOPQIBBgUrgQQAIwOBhgAE
ATv6duLhm/ZF3CPvDcqDDvGFi12tWLxLOeXTarYDeWoCMclQ7MKSDQXOzOVSsdpA
yLqqi72zHGhzpaTpEfiyqX++AIi4rpviDM7rcQYaKsUkQsXvr+aVUfV7BO8K51TL
wMh0p0e7jvVW1ivK4i0oya5LUQD0bYNJIxsbzNlFhGHVCGZkMAoGCCqGSM49BAMC
A4GLADCBhwJBbbj4YHDRaXBYv90YSFDnJsSnXzkLDGWw5mF7s26MhOOhuudrBb0J
pR8/fB3F3K7Z57KQcF4Cf2T/0aeAvfM4WawCQgERVN+j3qcG8PgfZgAKJPFj6J2C
4YclPZ9db4F+GOcbwG2fuBDH/FD5AGyPGaYC/J7tj4n5PIcwGpWbRiINoBkx+Q==
-----END CERTIFICATE-----
```

Generating the certificate on the fly would be best,
but I think that is ok to also use a pre-generated PEM,
and add as a comment the command used to generate the cert

for example

```
$ openssl req -x509 -newkey ec -pkeyopt ec_paramgen_curve:secp384r1 -keyout key.pem -out cert.pem -sha256 -days 3650 -nodes -subj "/"
$ cat cert.pm
```

**Review by adiroiban — CHANGES_REQUESTED** [review]
Thanks for the update.

Not sure what to say about ubuntu one cert.

It's good to have a certificate from real world, but for testing, I would prefer a certificate that is more focused on the scope of that particular test.

I  would argue that a barebone certificate is better here.


I left a few comments inline.

Thanks for the update.

**Review by adiroiban — APPROVED** [review]
Thanks for the PR.

I am not happy with using the Ubuntu One cert here... but I think this is good enough.

**Comment by adiroiban** [comment]
I have enabled auto-merge

**Pull request merged**